### DISCUSSION TITLE
Future Workflow

### DISCUSSION THREAD

**Discussion by bgregs514** [bgregs514/vscode-tm1#20] [General]
I expect this to be a heavily debated topic, since (realistically) it will redefine what a TM1 workflow should look like going forward.

Let's start by defining what the existing workflow looks like so we can better understand the pain points and barriers.  An existing TM1 workflow generally has these characteristics:
- Direct editing of a .pro or .rux file (object) through a 3rd party text editor
- Editing of a process or rule object through Architect
- Editing the .json content of a rule or process object using a front-end tool and posting/patching the result back to the server through the Rest API

The issue with all of these approaches is that they all try to update the TM1 object in a single event.  This leaves little room for git integration or alternative methods of development since there is no separation between development (traditional code file) and server object.

I thought about this quite a bit over the weekend and I think there may be a better alternative - splitting development into 2 steps.  Let's take a look at what this would look like:
- User downloads .json content of all TM1 code objects (rules and TIs) to a local workspace
- User edits these files like a traditional programming language
- User saves the file - "saves" in a traditional sense meaning it is saved to their local workspace and not to TM1 (yet)
- User has the option to use a 3rd party extension (layered on top of vscode-tm1) to push to a git repository, or push updated code content to the TM1 server object

The benefits of this approach exist solely in the pause between save and server commit.  Almost every other programming language code file can be copied down or cloned, edited, and then pushed back to a git repository.  The only thing that makes TM1 different in this sense is that the code needs to added to a valid TM1 object in order to function.  Why is pausing a good thing?  Let's take a slightly modified look at a better workflow using the above approach:
- User downloads .json content of all TM1 code objects (rules and TIs) to a local workspace
- User edits these files like a traditional programming language
- User saves the file - "saves" in a traditional sense meaning it is saved to their local workspace and not to TM1 (yet)
- User uploads code to a development TM1 instance for testing
- Once testing is completed, user uses 3rd party extension to push local file to a production (or user testing) git repository.

The pause allows more flexibility in where the code goes, because honestly that's the way it should be.  Gone (should be) the days of copying object files back and forth.  We have the ability to send just the important code pieces, so let's do that.  This effectively defines a full end-to-end workflow like such:
- User downloads .json content of all TM1 code objects (rules and TIs) to a local workspace
- User edits these files like a traditional programming language
- User saves the file - "saves" in a traditional sense meaning it is saved to their local workspace and not to TM1 (yet)
- User uploads code to a development TM1 instance for testing
- Once testing is completed, user uses 3rd party extension to push local file to a production (or user testing) git repository.
- TM1 Rest API git integration is used to deploy TM1 objects from a remote git repository

Notice how we are only directly pushing updates to a TM1 object in the development environment for testing, and we let git do what it does best for the rest of the workflow?  I think it works pretty well, and the pause in between allows us to do fancy things like "check for orphans", "diff servers/local workspaces", "roll back changes", etc. - most of which could easily be implemented through use of additional VSCode extensions.

What do you think?

**Comment by TM1Fanboy** [comment]
Sounds  good but why direct push to a test server? Couldn't we use a different git branch here?

**Reply by bgregs514** [reply]
Absolutely!  The way I'm thinking about it is to have a single "upload" icon next to each object which will push the code to the corresponding object on the TM1 server - this will most likely be a development environment.  This step is entirely optional however, and any other extension could be used to send the code to any git repository, or do anything with it really.

The only reason I thought about having a button that directly pushed code to an object was for development testing purposes.  I assumed that the code would need to be tested and debugged on a development environment before the user would want to commit to a repo, and this would just be a shortcut instead of having to commit to a dev repo, deploy from dev repo to dev model, etc.  That being said, this new approach would be completely flexible, so the "direct push" button could be excluded from the development process without any issue.

**Comment by ykud** [comment]
Hi @bgregs514 . 

I was rewatching Fargo yesterday, so I just have to get "I'm not going to debate you Jerry" out of my system :) 

With that done, here's my take on this, I was thinking about it for a bit in the past.

One big difference between traditional programming languages & TM1 is that you cannot run / compile / error check code locally, so the 'local code save' step is not as useful for TM1 development. 
The way I saw it was as following:
* You connect to a Tm1 server 
* Opening a rule or process
   *  saves the object files (there's actually a couple of them, I'll talk about it below) to the specified folder (workspace folder as you use now, but server specific one) in the same folder structure as Git export (i.e. with cubes / processes / etc subfolders) and same file extensions
   * (Optionally) compares the code in your repo to the what's on the server and says something along the line of 'this rule was updated since you last had a copy'
   * Opens the code file for editing
* You can edit the code (rule / TI) in the editor 
* Saving would do the following:
   * Save the file to folder
   * Update the code on server you're connected to check syntax / validate results (here's where we differ, I think it should be the same action, without a 'pause')

So you'd be developing pretty much as in Architect / PaW now, with saving code and checking results. The codebase copy to file system will be useful for:
* search functionality 
* IntelliSense 
* if using git -- this folder will be your local 'repo' that you can push from to another environment or do any comparisons. 

I'm not convinced that the local code update - pause - upload to server step brings much to the table, I'd rather have save directly update the objects on the server you're connected to (probably dev environment) and leave it open whether people would want use git to promote things further in the chain. 

Coming back to multiple files per object, there are 2 files for cubes and processes:
* json file with cube structure or process definition with variables and source type, etc
* code portion for rules / TI code

both are important as you cannot create a new TI without specifying both, but also can be useful if we ever get fancy enough to look into a bit more 'model-aware' IntelliSense as in you can automatically suggest what parameters TI has for RunProcess or what dimensions are there in a cube for CellPutN.

**Reply by bgregs514** [reply]
Hahaha!  I think you just inspired me to rewatch Fargo! 😄

To start off, thank you for that overview - reading through your response helped me rethink certain parts of my original write-up.  I'm in agreement with you, I don't think I had fully thought through the "pause" piece.  My original hesitation was that chaining too many events (save to server and save to file) together would give the user less freedom over workflow, but you're right, adding the "save to server" part as a separate event would be more painful for users.  I'd say 99% of the time the code will need to be pushed to the server at the time of saving, so let's add that into the main save event.

Here's the part I'm still hung up on: VSCode doesn't let you name tabs.  It's a read only attribute, so the only way to name that is by opening a named file.  Initially I thought, no big deal, we're already saving local copies, but if I recall correctly, the way TM1 saves rule and TI files is in .json format.  This means that if we save them as .json files also, the _only_ way to determine what language file to present is either:
a) We look at which sub-directory the file was in - i.e. did it come from the "cube" folder or the "process" folder
b) We attempt to determine which TreeView the element was in and assign the syntax file based off of that - this unfortunately isn't as easy as it sounds based off of my initial testing, but I could be missing something.  Not to mention it would mean that we would need to populate the TreeItems off of the server, instead of reading from a local directory which is much faster regarding startup times

Even if we did determine the language on the fly (which I am fully confident we can do this), we would basically be giving up any type of auto-detection based on file extension - something where VSCode really excels.  Not to mention we would need to develop the additional code to handle it - not a big deal, but worth noting.

That leads me to my workaround - granted this is just an idea and I'm not really thrilled with every aspect:
1) When a local workspace is initialized and the files are saved from TM1, we save them in 2 formats: the normal TM1 .json structure, and our own with meaningful extension names
2) When the TreeViews are populated, this becomes trivial and very natural for VSCode to populate elements based off of matching extension names (all of the "path" and "fs" modules support all of these functions) - not to mention we only have to poll the server once at startup (when it downloads names and code contents), and not on every `onSelectionChange` event
3) Language matching is completely handled by the extension names we include in the package.json file - __no coding required__ 🥳
4) This is where it's slightly messy (but not too bad) - when a file is saved, it commits to TM1 (easy, we already have this), it saves to our custom extension file (no action necessary here), it saves to the corresponding .json file in the natural TM1 structure (this should be easily handled by the "path" and "fs" modules).

This would basically mean that a folder structure in the natural TM1 git format would be a bi-product of our extension, and not the main working directory.  An important note when it comes to moving files up using another git extension.

Now I'm not 100% sold on this idea yet, but my initial reaction is that having meaningful extension names greatly outweighs the inconvenience of having to save 2 copies of the file.  Even in the largest models, we're only talking about the code portion of TIs and rules, so file sizes shouldn't exceed ~5mb at most I would think.

The benefits would be that we let VSCode automatically handle the language assignment, and the tab names become more meaningful (instead of everything being postfixed with .json).  Imagine if a rule file and a process were named the same, the icon wouldn't display in the tab, so the only way to determine the difference without clicking into the tab would be based on extension.  Saving to the correct .json format during the main save event isn't as clean as I would like, but overall I think the benefits outweigh it here.

I should have a working example of this proposal on a separate branch uploaded shortly.  I'm using the following naming convention for extensions, but we can change it:
- rule = .rul
- control cube rule = .rulx
- process = .pro
- control process = .prox

__All this being said, the average user won't notice any impact other than the "git" ready directory existing in a different location.__

__EDIT:__
I just thought of another major benefit to having custom extension names.  We currently are forced to track the type, name, and document handle of each file opened in the `g_OpenDocuments` variable.  Some of the dependency on this is removed once tabs are named (the "name" piece isn't required), but everything having a .json file would mean we would still need to determine and track "type".  Including meaningful extensions removes this dependency since the "type" can be directly inferred from the name of the tab.  The document handle would also go away as soon as we are dealing with real files, but this will be the case for both the .json or custom extension.  Being able to determine type from file name is just __so__ much easier instead of trying to determine it on the fly.

The extensions don't have to be displayed in the TreeViews (and I don't think we should show them), but the tab itself will include the extension.  Getting off of the global Document variable would be a major win for us in terms of flexibility and scalability.

I think I sold myself on this idea - please let me know if there are any issues that I'm not seeing. 😄

**Reply by ykud** [reply]
Hi @bgregs514 ,

> Here's the part I'm still hung up on: VSCode doesn't let you name tabs. It's a read only attribute, so the only way to name that is by opening a named file. Initially I thought, no big deal, we're already saving local copies, but if I recall correctly, the way TM1 saves rule and TI files is in .json format. 

It might be all a lot easier, I've opened a planning sample git repo I played with to illustrate how git export looks: 
https://github.com/ykud/git_tm1_planning_sample

So for cubes you have 2 files:
* a cube_name.json file with structure and dimension lists: https://github.com/ykud/git_tm1_planning_sample/blob/master/cubes/plan_BudgetPlan.json 
* a cube_name.**rules** file with rule contents that match select=Rules without any transformations: https://github.com/ykud/git_tm1_planning_sample/blob/master/cubes/plan_BudgetPlan.rules

For processes you have 2 files:
* a ti_name.json file with source type / parameters and variables : https://github.com/ykud/git_tm1_planning_sample/blob/master/processes/mover.json
* a ti_name.**ti** file with TI code, that is exactly the select=Code output : https://github.com/ykud/git_tm1_planning_sample/blob/master/processes/mover.ti

So from my point of view by having syntax highlighting match 2 file extensions (ti & rules) you'd get what you want without additional file to maintain, as well as decent enough document tab names?

**Reply by bgregs514** [reply]
You are exactly right!  I was completely mistaken here - I guess that's what I get for being too lazy to git init a model prior to writing that up. 😉

I had a good talk with Hubert, and he helped clarify a few things for me around git integration.  Even if we did use the TM1 git repo (.json/.rule/.ti files) as our source, there is no guarantee that those match what is currently running in the instance.  If I understand correctly from the conversation, tm1 git was intended to copy the current state of a given model, and then deploy it up to a higher environment.

In this aspect it differs from a typical workflow (our project for example), where we develop, push to git, and then run the final product.  In the tm1 world, we develop, push to an instance, snapshot the _entire_ instance to git, and then use the tm1project file to deploy to other environments.

My concern would be that if we use a tm1 git repo as the source, it has a high possibility of being out of sync with the running instance - "high" because I'm assuming the current adoption rate of users who actually deploy a model off of git is low since it would involve changing their previously existing workflow.  This is just anecdotal though, so please prove me wrong. 😄

The other issue with a "selective" git commit if we did decide to go down that path (i.e. committing normally to a git branch for individual files) is that TM1 doesn't support it internally.  If my understanding is correct from the conversation I had, doing a commit outside of the TM1 git framework would overwrite any previously tracked history of commits - that would be a problem.  I'll take this point as a follow up, because I don't see how we would work around that one.

The other concern is deploying frequently for larger models.  If we did get around some of these issues and begin to use git as the method to push individual rules and TIs, would this have any impact on the larger models?

Anyway, I'm not saying it's not possible (in fact I'm all for traditional git integration), but I think the easiest path to a v1 release would be custom files that contain the source of the __running model__, and potentially some high level commands to initiate an internal tm1 git commit and pull plan - possibly with some auto-assisted editing of the tm1project file too.

Just as a clarification, these generated files that we have would never be known to the typical user.  In fact, I just dump them all into a single directory currently and use VSCode to piece them out to the right place based off of extension.  Once the file is created, it's relatively low maintenance on our side.  That being said, it is another moving part and I'm happy to be rid of them if we can use something else.

Sorry for yet another rambling post - I do like how this is self-documenting, but I'm also open to setting up some meetings if you prefer that route. 😄

**Reply by ykud** [reply]
Hi @bgregs514 , 

This is all getting quite confusing because we're trying to solve 2 things at once and I think they are better off tackled separately :) 

The way I see it we have 2 topics here:

1) VSCode wants to have files stored for tab names

I think the solution to this is to store model code in the same structure and file names as git integration does (including subfolders) to make it searchable.  Despite this being related to 'git integration', it does not necessitate using git. 
The workflow is 'click on an object - extension downloads it as a file - edit - press save - extension saves a file and updates the code on server', same as if you'd be doing it in Architect or PaW with an added benefit of VSCode plugin libraries, formatting and searching model code. 

This is what I think is the goal of this extension in v1 :) 

> I just dump them all into a single directory currently and use VSCode to piece them out to the right place based off of extension. 

I see benefit in maintaining folder structure and file extensions in  the same was as git integration does (with cubes / processes / chores) if only for cleanliness and not having to invent new formats. 

2) How to use git in tm1 model development and how git fits into picture. 

I think you're going down the familiar rabbit hole of 'how to use git in tm1 development' and it's a deep one, been there :)  This is a large topic for discussion (I wrote a bit on it already) ), but in my opinion it is kinda 'orthogonal' to how extension should work.  The main reason I want to have a folder / file extension structure similar to git is to 'enable' repo compare, not to 'force' git's usage.  Anything git-related should be v2 version of the extension.

> In this aspect it differs from a typical workflow (our project for example), where we develop, push to git, and then run the final product. 

I'm pretty sure that for this project you develop **and compile and test** extension locally and only when you're happy with the result **push to git**. 

The 'save button uploads code to server' will mean that TM1 server you connected (the running instance) acts as place where you **compile and test** changes prior to **pushing them to repo**.

> The other issue with a "selective" git commit if we did decide to go down that path (i.e. committing normally to a git branch for individual files) is that TM1 doesn't support it internally. 

Funny enough it does, just in a very awkward fashion. You can create tm1project file that ignores everything but the few objects you want to promote and push it to a branch -- it'll push only unignored objects and pulling from that branch will get you only them. 

>  I do like how this is self-documenting, but I'm also open to setting up some meetings if you prefer that route. 😄

Sure, I think we have a bit of timezone overlap in your evening / my morning (I'm in Au), happy to have a chat, can be a bit faster than typing another git-rant :) 

Cheers,
Yuri

**Reply by bgregs514** [reply]
@ykud Apologies for the delay in getting back to you!  I think I'm in agreement with all of your points - let me take a stab at restructuring the current local workspace layout to match the TM1 one.  I pushed out a few updates to get TI running/creation incorporated into the extension, so I'm finally able to get back around to working on polishing the local workspace structure.

Once this is in and I'm able to polish a few other areas of the extension, I may reach out for a quick meeting just to go over current functionality and scope for a v1 release.  I think we are getting closer!  Thank you again for all of your help. 😄

**Comment by bhushpar** [comment]
Hi @bgregs514 

End-to-End Tm1 integration is 3 stepped process - 
1) Bringing TM1 Objects from (environment - test/uat) to your local git directory
2) Performing differential with repo and commit
3) Integration to Organization's pipeline to perform live migration to the server( env - uat/prod)

Automation around Pt1:
One thing to keep in mind is when you performing pull - you not only pull tm1 objects - dim,rules,process,attr cubes, etc. but also non tm1 objects like scripts- py, vbs, etc. You might also want to consider connection type->  server type: Windows or Linux - both have different connection protocol. 
Ideal workflow:
Develop on Server using - Architect or any 3rd party tool -> Validate development (rules/TIs/py) -> pull to local -> create diff with repo(uat/prod) -> commit to repo(uat/prod) -> Live migrate to (uat/prod using deployment manager)

P2: will be pretty straight forward via git (would encourage to use git and not do any automation here - works like a charm)

P3: this depends and may vary by organization - but will require scripting to create pipelines with perform live migrations to uat/prod using deployment managers like - Jenkins.

Bhushan

**Comment by BenStruwe** [comment]
Hi @bgregs514,

another thing to consider is, what makes up a repository for a server.
Surely there are shared objects e.g. TIs like }bedrock that are part of a server but should ideally be maintained in a separate package. 

A deploy-pipeline should handle such dependencies and push to various servers. 

With regards to the general process flow, I agree to allow online changes within TM1 on dev directly and have the source code handled via git including deployments to test/prod.

Cheers,
Ben

**Reply by bgregs514** [reply]
Hi @BenStruwe,

That's a very interesting thought.  Almost like including submodules from other projects into the main tm1 git repo - wow, that has the potential to be really powerful!

In its current state, I'm not yet sure if TM1 can handle selective commits done outside of its own git api, but as soon as we are able to figure out the possibilities/limitations of it, we should definitely look into this.

Thank you for the feedback!

**Reply by TM1Fanboy** [reply]
I am not sure I understand your problem here @bgregs514. You can easily download the git repository containing the files to your local system, change something, push it back. The only thing you have to do afterwards is initiate a pull to the specific tm1 instance. I did that for example to test hot deployment of new bedrock processes.
 
I love the idea of @BenStruwe. That's were my above mentioned tests came from. I would orchestrate the whole dependency stuff git-wise. For example always pulling certain 'modules' not from a tm1 server but from a centralized git. That way you always get a new version if available and automatically deploy that to your tm1 instance. I think in git ('normal' git, nothing tm1 specific) all the things you need for that are already there.

**Reply by bgregs514** [reply]
@scrumthing I fully agree with that approach.  In all honesty, if we were working from a git repo as the source, we could accomplish almost every task with existing extensions - git, rest api call, syntax highlighters, etc. already exist for all of the pieces we need.  A case could be made for a very light extension that encapsulates some of the more involved rest requests, but that wouldn't be too difficult.

The problem that I see is that there are only a small amount of full implementations that have baked the git workflow into their models.  I would venture to say that most deployments currently consist of either some rest operation patching/posting updated objects to a live model, or a manual copy of the objects and a server restart.  I don't think I have yet seen a full blown implementation rely solely on git to run deployments.

The issue that I think we're going to run into is that __if__ git has been initialized on a model, there is still no guarantee that it matches the code on the running instance.  My understanding from a handful of meetings is that tm1 git is not intended to be directly modified.  It is there to snapshot a given instance at a point in time, and then deploy that instance to other tiers of the model.  That effectively switches the true source of the code/data from being a git repo to a live model.

Compound that with the fact that I don't think most models are running off of a tm1 git repo, and I think we are going to have issues with adoption of our tool if we try to make the workflow use a tm1 git repo as the source of the code.

Now, all that being said, I agree that having a git repo as our source is something we should strive for, I just don't think we are there yet.  If we do see more models transition to updating git source, and then pushing that out to a live instance instead of the current "live model -> git" workflow, then I think we can consider switching some things around.

Anyway, I'm still stumbling my way around many of these topics, so please correct me!  I have learned more in these discussions than I have tinkering around on my own. 😄

**Comment by VadymS77** [comment]
Hi everyone
The idea of a pause between development and deployment to a model or branch in Git via VSCode is quite clear and, for me, it's a great idea. It is used in other programming languages with  integration with Version Control System.
But how do you see in a perspective the process of rules creation via VSCode, very often you need to see a result  of calculation in a cube immediately.
It turns out that I need create rule, for example, in the Architect, and CheckIn or CheckOut to the required branch from VSCode using  this extension or directly using TM1-Git RestApi.
+ maybe I have an old version, but for example, when you triple click on an object, three identical tabs open.

Vadym

**Reply by bgregs514** [reply]
Hi @VadymS77,

Thank you for the feedback!  Rule/TI creation is a very interesting topic, and one that I don't think we've quite nailed down just yet.  My initial reaction would be that everything needs to be created on a running instance first, and then the extension can be used to create a tm1 git plan to commit the model to a remote git repo.

For [reasons](https://github.com/bgregs514/vscode-tm1/discussions/20#discussioncomment-498163) described in my comment above to @ykud, I'm not entirely sure if we can do selective commits directly to a tm1 git repo.  My understanding is that tm1 git integration is slightly inverted when compared to a traditional workflow - it's designed to develop, push to a running instance, snapshot the entire model at a point in time, and push that to a git repo.  That slight deviation has a few pain points for us currently.

In regards to the issue you are experiencing, that does sound like an older version.  I won't go as far as to say that the current build is "stable", but now that it's reliant on local files it feels much more natural when editing files. 😄

**Comment by Bilel-JALLOULI** [comment]
Hello,
I have the error below when i click on the "add new connection button"

[2021-04-21 18:50:52.969] [renderer1] [error] command 'vscode-tm1.openAddConnectionScreen' not found: Error: command 'vscode-tm1.openAddConnectionScreen' not found
    at w._tryExecuteCommand (file:///C:/Users/33669/AppData/Local/Programs/Microsoft VS Code/resources/app/out/vs/workbench/workbench.desktop.main.js:1558:3364)
    at file:///C:/Users/33669/AppData/Local/Programs/Microsoft VS Code/resources/app/out/vs/workbench/workbench.desktop.main.js:1558:3245
    at async u.open (file:///C:/Users/33669/AppData/Local/Programs/Microsoft VS Code/resources/app/out/vs/workbench/workbench.desktop.main.js:627:18599)
    at async l.open (file:///C:/Users/33669/AppData/Local/Programs/Microsoft VS Code/resources/app/out/vs/workbench/workbench.desktop.main.js:627:20623)
